SQLite MessageIndex

silverailscolo opened the pull request:
Migrating from storing and retrieving messages using as index a set of nested Dicts to a SQLite central messages index (See our "[Untangling the Organic Knot](https://github.com/ramses-rf/ramses_rf/discussions/191)" discussion).

Stores ~the _parsed_ message payload as JSON in column `pl`, plus~ a summary of all enclosed keys in column `plk` for checking before loading the payload itself. Eventually replaces all uses of `_msgz_`.

TODO:

- [x] @zxdavb Confirm this is the intended approach
- [x] Test fetching payload FAN_INFO values for my Vasco
- [x] Add a test for this process (current tests keep passing, so... they don't fetch?)
- [x] Use this table instead of the ~`_msgs_` and~ `_msgsz_` Dict
- [ ] Fix any issues
- [x] Remove unused `_msgz_` Dict from `ramses_rf/entity_base.py` and `ramses_rf/database.py`

silverailscolo committed:
rename _zzz not private, deprecate Dicts

silverailscolo committed:
async locked housekeeping

silverailscolo committed:
First query

silverailscolo committed:
Refine query

silverailscolo committed:
_31D9 query example

silverailscolo committed:
Add hvac test, cpmplex Code, add element required for reverse testing

silverailscolo committed:
fix mypy missing return

silverailscolo committed:
use sql query in _msg_value_code

silverailscolo committed:
use sql query in _msg_value_code

silverailscolo committed:
use sql query in _msg_value_code

silverailscolo committed:
comment out

silverailscolo committed:
comment out

silverailscolo committed:
comment out

silverailscolo committed:
MessageIndex query typed td, add docstr

silverailscolo committed:
tweak database sql, activate 31DA msg reverse test

silverailscolo committed:
migrate _msgz to msg_db, tests

silverailscolo committed:
fix lint, shorten contains()

silverailscolo commented on src/ramses_rf/database.py:
```diff
@@ -199,12 +243,16 @@ def _insert_into(self, msg: Message) -> Message | None:
                 msg._pkt._ctx,
                 msg._pkt._hdr,
+                json.dumps(msg.payload, indent=4),
+                payload_keys(msg.payload),
```
Took out the JSON, pretty clear

silverailscolo committed:
fix tuple lint

silverailscolo committed:
require key in sql, add msg_db assert (not run?)

silverailscolo committed:
qry dev._gwy.msg_db

silverailscolo committed:
fix qry result

silverailscolo committed:
refine qry result

silverailscolo committed:
add value_qry

silverailscolo committed:
fix entity_base mypy

silverailscolo committed:
don't add None keys to plk index field

silverailscolo committed:
activate, start msg_db, contains service query, move ventilator test to tst/_rf/device add msg_db

silverailscolo committed:
rename entity_bas._msg_db to _msg_list

silverailscolo committed:
fix some code for tests, turn off a test to investigate

wimpie70 commented:
I'm getting errors like:
2025-10-03 12:18:39.905 ERROR (MainThread) [homeassistant.helpers.dispatcher] Exception in add_devices when dispatching 'ramses_cc_new_devices_sensor': ([37:126776 (CO2), 37:168270 (REM), 37:154519 (HVC), 32:153289 (FAN), 29:176861 (REM), 37:169161 (DIS)],)
Traceback (most recent call last):
  File "/config/custom_components/ramses_cc/sensor.py", line 121, in add_devices
    and hasattr(device, description.ramses_rf_attr)
        ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/ramses_rf/device/hvac.py", line 958, in fan_info
    res_mode: list = self._msg_qry(sql)
                     ~~~~~~~~~~~~~^^^^^
  File "/usr/local/lib/python3.13/site-packages/ramses_rf/entity_base.py", line 568, in _msg_qry
    for rec in self._gwy.msg_db.qry_field(sql, (self.id[:9], self.id[:9])):
               ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/ramses_rf/database.py", line 419, in qry_field
    self._cu.execute(sql, parameters)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
sqlite3.OperationalError: near "%": syntax error

silverailscolo commented:
That, and 
`src/ramses_rf/entity_base.py:294: error: Missing return statement  [return]`
Working on it, in steps

silverailscolo commented:
@wimpie70 Your `2411` tests pose a new problem with the SQLite gwy.MessageIndex:

At the end of `mock_gateway()`, we need to async start the SQLite MessageIndex. It takes a little bit of time, including housekeeping to throw out "old" messages. Could be AsyncGeneraor or the like.
Please have a look and tweak that part.

```
@pytest.fixture
def mock_gateway() -> Generator[MagicMock, None, None]:
    """Create a mock Gateway instance for testing."""
    gateway = MagicMock(spec=Gateway)
    gateway.send_cmd = AsyncMock()
    gateway.dispatcher = MagicMock()
    gateway.dispatcher.send = MagicMock()

    # Add required attributes
    gateway.config = MagicMock()
    gateway.config.disable_discovery = False
    gateway.config.enable_eavesdrop = False
    gateway._loop = MagicMock()
    gateway._loop.call_soon = MagicMock()
    gateway._loop.call_later = MagicMock()
    gateway._loop.time = MagicMock(return_value=0.0)
    gateway._include = {}
    # Add msg_db attribute that's accessed by the message store
    # gateway.msg_db = MagicMock()
    # gateway.msg_db.get.return_value = {}. < this is now in master
    # activate the SQLite MessageIndex
    gateway.msg_db = MessageIndex(). # < we need this to get _msgs()

    yield gateway
```
Also, the device's self._msgb() dict[code, msg] is rebuilt from the MessageIndex every time it is called, in entity_base.py#_msgs(self) . So adding stuff to it in `handle_message()` won't work.

Also, the MessageIndex has a fixed size of 4 chars for the code (as str), breaking on your extended `2411_dev` keys.
Using SQLite queries we can perhaps get to the same stuff. For example, there is a `plk` field that holds all (non-None) payload keywords, separated by `|`.
Or perhaps in the tests you can use only the _msgb_ dict.

silverailscolo committed:
fix qry LIKE, strip _msgz_, add code notes

silverailscolo committed:
msg_list from qry

silverailscolo committed:
tweak qry format

silverailscolo committed:
add dtm to debug log contrived

silverailscolo committed:
annotate, add debug, handle ctx

wimpie70 commented:
Thought you wanted to do as much as possible in the database... but yes sure, a dict will work, and store it in hvac ? Or at another lvl. Where should it be available ?
(Builder Pattern)
After restart or empty cache, hvac will send a discovery RQ for 1 param. If we get any 2411 I or RP from a fan, it will get_all_parameters ( or update_? Not at my compu), execute a callback and create or update entities in _cc


⁣BlueMail voor Android downloaden ​

Op 6 okt 2025, 13:13, om 13:13, Egbert Broerse ***@***.***> schreef:
>silverailscolo left a comment (ramses-rf/ramses_rf#196)
>
>> a special table for 2411 or maybe better
>
>Could a good ol' dict work for that? If you use the code as key, and
>the value as such, you'd have all values available.
>What would happen after reboot? The cached logs will be interpreted,
>but that would mean only a single 2411 msg.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/ramses-rf/ramses_rf/pull/196#issuecomment-3371089500
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

silverailscolo committed:
add msg_db unittest

silverailscolo committed:
extend db test

silverailscolo committed:
clean up, remove _msgz_, database(maintain) option for 2411 tests

silverailscolo commented:
@wimpie70 The way I understand your need to keep multiple 2411 statuses, a plain dict[code, value] can do a lot in python. No setup required.
The power of the SQLite MessageIndex will be of use to gain custom access to payload key:values across Codes. See the "co2_level" demo in hvac.py. Current handling of 4-5 nested msg dicts, especially for a wider range of applications, has reached the limits of dicts (David told me).
Just fixed your `2411` mocked tests by adding a `maintain` parameter to the MessageIndex init. Default is True for daily cleanup of stale msg, not important in tests unless that is what you test ;-)
Also added a unites for MessageIndex. Found that really useful to understand the inner workings. Still a couple of tests to fix.

silverailscolo committed:
Merge branch 'master' into eb-sqlite-messagebase

silverailscolo committed:
refactor qry, fix mypy

silverailscolo committed:
skip 2 2411 tests, mark fix

Commit:
2411 params now use a dict to store values in the device

Commit:
fixed test_hvac_ventilator to match the changes in hvac.py

Commit:
fix normalization  in hvac._handle_2411_message

silverailscolo committed:
Update requirements_dev.txt

silverailscolo committed:
pull master, 2411, fewer vars in tests
